crates/db_schema/src/impls/community.rs

```diff
@@ -209,23 +210,62 @@ impl Community {
   pub async fn get_random_community_id(
     pool: &mut DbPool<'_>,
     type_: &Option<ListingType>,
+    show_nsfw: Option<bool>,
   ) -> Result<CommunityId, Error> {
     let conn = &mut get_conn(pool).await?;
 
-    let mut query = community::table
-      .filter(not(community::deleted))
-      .filter(not(community::removed))
-      .into_boxed();
+    // This is based on the random page selection algorithm in MediaWiki. It assigns a random number
+    // X to each item. To pick a random one, it generates a random number Y and gets the item with
+    // the lowest X value where X >= Y.
+    //
+    // https://phabricator.wikimedia.org/source/mediawiki/browse/master/includes/specials/SpecialRandomPage.php;763c5f084101676ab1bc52862e1ffbd24585a365
+    //
+    // The difference is we also regenerate the item's assigned number when the item is picked.
+    // Without this, items would have permanent variations in the probability of being picked.
+    // Additionally, in each group of multiple items that are assigned the same random number (a
+    // more likely occurence with `smallint`), there would be only one item that ever gets
+    // picked.
+
+    let try_pick = || {
```

Comment:
I doubt this needs to be a closure, its not going to get called anyway until the query is awaited. Not a big deal tho.

Reply:
It's called twice, and it returns a boxed query, which isn't clonable